### ISSUE TITLE
Accessibility: https://ipython.org/usersurvey2013.html

### ISSUE THREAD

**Issue by marv-tech** [jupyterhub/outreachy#88]
## What page is this for?

https://ipython.org/usersurvey2013.html

### Report Summary

1. **Error (1):** Language missing or invalid
2. **Contrast Errors (35):** Very low contrast
3. **Alerts (53):** This includes issues with Justified texts, suspicious alternative text and image titles.
4. **Features (3):** This had to do with images and their alternative texts.
5. **Structural Element (30)**
6. **Aria (1)**

<img width="1421" alt="image" src="https://user-images.githubusercontent.com/57637620/196569998-3f20c85c-c037-4178-a5a8-ac468cacadff.png">

### 1. Language missing or invalid error.
Using the wave software, the first error discovered was one relating to the failure to identify the language for the page. Because the language was not found in the codebase, wave flagged it. The code block below highlights the code statements that should include the lang (language) attribute as well as the accompanying image provided by Wave.

<img width="628" alt="image" src="https://user-images.githubusercontent.com/57637620/197022692-471b78cc-7a0e-4ae9-8ac3-3e593b02229e.png">

```html

<html xmlns="http://www.w3.org/1999/xhtml" style="margin-left: 380px;">
<head>
<meta http-equiv="X-UA-Compatible" content="IE=Edge">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
```
### 2. Contrast Errors
Contrast errors occur when there is very low contrast between the text and background making it difficult for individuals with low vision. Web Content Accessibility Guidelines(WCAG) necessitate that page elements have defined foreground AND background colours that provide adequate contrast. The average requirement is **4.50:1** and above, which the flagged contents didn't meet. There were 35 contrast errors in total and I have grouped them according to their current contrast ratios.

#### a. Contrast ratio of 4.07:1
As shown in the image, these are links found at the top of the page, and each of these code blocks displays the codes for each link in this category.

<img width="1023" alt="image" src="https://user-images.githubusercontent.com/57637620/197019233-5c9ca3aa-1988-49a9-8175-4f9b8df69378.png">

**Install Link**
```html
<a href="https://jupyter.readthedocs.io/en/latest/install.html" target="_blank" style="color: rgb(206, 92, 0); background-color: rgb(255, 255, 255);">
Install
</a>
```

**Documentation Link**
```html
<a href="documentation.html" style="color: rgb(206, 92, 0); background-color: rgb(255, 255, 255);">
Documentation
</a>
```

**Project Link**
```html
<a href="project.html">
Project
</a>
```
**Jupyter Link**
```html
<a href="https://jupyter.org/" target="_blank">
Jupyter
</a>
```

**News Link**
```html
<a href="news.html">
News
</a>
```

**Cite Link**
```html
<a href="citing.html">
Cite
</a>
```

**Donate Link**
```html
<a href="donate.html">
Donate
</a>
```

**Books Link**
```html
<a href="books.html">
Books
</a>
```
<img width="267" alt="image" src="https://user-images.githubusercontent.com/57637620/197019444-e43c58e1-b816-4172-b0d8-171c08e31cc3.png">

**Stack Overflow Link**
```html
<a class="reference external" href="http://stackoverflow.com/questions/tagged/ipython">
Stack Overflow
</a>
```
**Mailing list Link**
```html
<a class="reference external" href="https://mail.python.org/mailman/listinfo/ipython-dev">
Mailing list
</a>
```

**File a bug Link**
```html
<a class="reference external" href="https://github.com/ipython/ipython/issues">
File a bug
</a>
```
**Reddit Link**
```html
<a class="reference external" href="http://www.reddit.com/r/IPython">
Reddit
</a>
```

<img width="267" alt="image" src="https://user-images.githubusercontent.com/57637620/197019570-5054344a-b483-4c7a-824d-bfdc43030774.png">

**Development information Link**
```html
<a class="reference external" href="https://github.com/ipython/ipython/wiki/Dev:-Index">
Development information
</a>
```
**Travis CI Link**
```html
<a class="reference external" href="http://travis-ci.org/#!/ipython/ipython">
Travis CI
</a>
```

**Wiki Link**
```html
<a class="reference external" href="https://github.com/ipython/ipython/wiki">
Wiki
</a>
```
<img width="725" alt="image" src="https://user-images.githubusercontent.com/57637620/197019707-e8ad8de4-2dc1-4cee-aeb2-e0d857c07855.png">

**Spyder IDE Link**
```html
<a class="reference external" href="http://code.google.com/p/spyderlib/">
Spyder IDE
</a>
```
<img width="809" alt="image" src="https://user-images.githubusercontent.com/57637620/197019864-39d042b8-a29f-4b72-aa1e-83df94dbba44.png">

**Google Spreadsheet**
```html
<a class="reference external" href="https://docs.google.com/spreadsheet/pub?key=0AqIElKUDQl8tdHF2WmlKdTZTRlZVRGFGTDgtUXFBVUE&amp;output=html" style="color: rgb(206, 92, 0); background-color: rgb(255, 255, 255);">
```
#### b. Contrast ratio of 3.48:1

<img width="689" alt="image" src="https://user-images.githubusercontent.com/57637620/197028832-74754241-8ed2-4ab8-b675-c4f5de49a155.png">

```css
/* Footer */
div.footer, div.footer a {
  color: #888a85;
}
```
### 3. Alerts Errors
Alerts were issues flagged by Wave that could be an issue with accessibility; using the Wave software, a total of 53 alerts were discovered and divided into three groups. 42 of the issues involved justified text indicating the presence of fully justified text. Wave also detected 9 issues with image alt text and two images with title alerts, which indicate that an image has a title attribute value but no alt value.

#### a. Justified text

```css
div.body {
  padding-right: 2em;
  text-align: justify;
}
```
<img width="689" alt="image" src="https://user-images.githubusercontent.com/57637620/197033980-66b8327e-334f-4fed-8b49-47da20e85463.png">

#### b. Suspicious alternative text
Alternative text is likely insufficient or contains extraneous information.

`<img class="logo" src="_static/IPy_header.png" alt="Logo">`

`<img alt="_images/countries1.png" src="_images/countries1.png" style="width: 697.5px; height: 459px;">
`

`<img alt="_images/ipy_versions.png" src="_images/ipy_versions.png">`

`<img alt="_images/platforms1.png" src="_images/platforms1.png">`

`<img alt="_images/py_versions.png" src="_images/py_versions.png">`

`<img alt="_images/installation.png" src="_images/installation.png">`
 
`<img alt="_images/help_resources.png" src="_images/help_resources.png">`

`<img alt="_images/role.png" src="_images/role.png">
`
#### c. Image with a title but no alt attributes
This occurs when an image has a title attribute value but no alt attribute value.

<img width="251" alt="image" src="https://user-images.githubusercontent.com/57637620/197041571-fd93844e-5d7a-420e-b724-afa97daddc32.png">

`<img src="_static/nbviewer-thm-9.png" title="IPython Notebook Viewer">`

<img width="251" alt="image" src="https://user-images.githubusercontent.com/57637620/197041985-c436dc2d-fe55-47ff-a463-47d60d15170d.png">

`<img title="Packt Publishing donates a portion of the proceeds from this book to support IPython's development." width="160" height="200" id="sidebar-book-cover" style="margin-bottom: 10px;" src="_static/ipython-cookbook-2nd.png">`

### 3. Features
These points indicate areas where an element if properly implemented, will improve accessibility. "Linked image with alternative text" was used to highlight three issues. These three issues emphasised the proper use of alternative text while flagging the contents of these alternative texts.

<img width="714" alt="image" src="https://user-images.githubusercontent.com/57637620/197053137-a5b409d0-a6a6-4e69-90e8-756c53b43fc8.png">

`<img class="logo" src="_static/IPy_header.png" alt="Logo">`

<img width="476" alt="image" src="https://user-images.githubusercontent.com/57637620/197053221-b46a64df-4f38-4540-a507-a9da968c9f76.png">

`<img src="_static/jupytercon-logo.svg" width="180px" alt="JupyterCon 2017">`

<img width="762" alt="image" src="https://user-images.githubusercontent.com/57637620/197053457-5e41be6a-66a2-4de0-afec-5af7738a4683.png">

`<img alt="_images/countries1.png" src="_images/countries1.png" style="width: 697.5px; height: 459.0px;">`

### 1. The solution to Missing or invalid lang attribute value.
This can be resolved by identifying the document language using the <html lang> attribute with a valid value (e.g., <html lang="en"> for the English language).
```html

<html xmlns="http://www.w3.org/1999/xhtml" style="margin-left: 380px;">
<html lang="en">
<head>
<meta http-equiv="X-UA-Compatible" content="IE=Edge">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
```

#### a. Adjusting the contrast ratio to 7.06:1 will make it easier to resolve areas that had a contrast ratio of 4.07:1. This entails changing the colours of these elements to the suggested variant of that specific colour, as indicated in the CSS code below.

```css
div.header div.rel a {
  color: #8F4300;
  letter-spacing: .05em;
  font-weight: bold;

```

#### b. As previously stated, the contrast ratio of the footer elements was 3.48:1, which will be fixed by increasing it to 7.04:1. Increasing the contrast ratio necessitates changing the colour to #595954, as shown in the CSS code snippet below.

```css
/* Footer */

div.footer, div.footer a {
  color: #595954;
}
```
#### a. To fix this, change the text-align to `text-align: left;` or remove it entirely to use the default alignment.

```css
div.body {
  padding-right: 2em;
  text-align: left;
}
```
#### b. The solution to suspicious alternative text.
To assist screen readers, changing the alternative text (alt) to a text that conveys the actual content of the image and also very informative is the solution.

The alt text can be updated to **IPython Interactive Computing Logo**
`<img class="logo" src="_static/IPy_header.png" alt="IPython Interactive Computing Logo">`

**The alt text can be updated to "A chart of survey Respondents by country"**
`<img alt="A chart of survey Respondents by country" src="_images/countries1.png" style="width: 697.5px; height: 459px;">`

**A survey of respondents' preferred Ipython versions**
`<img alt="A survey of respondents' preferred Ipython versions" src="_images/ipy_versions.png">`

**A survey of Respondents' preferred installation platforms**

`<img alt="A survey of  Respondents' preferred installation platforms" src="_images/platforms1.png">`

**A survey of respondents' preferred installation methods**
`<img alt="A survey of  respondents preferred installation methods" src="_images/installation.png">
`
**A survey of respondents' preferred support resources**
`<img alt="A survey of  respondents preferred support resources" src="_images/help_resources.png">`
 
A survey of  Ipython respondent's roles
`<img alt="A survey of  Ipython respondents roles" src="_images/role.png">
`
#### c. The solution to Image with a title but no alt attributes
To solve this issue, the inclusion of alt attribute to the images will resolve this and make it more accessible.

**Comment by minrk** [comment]
Thank you for the detailed report and excellent suggestions!

**Comment by marv-tech** [comment]
Thank you @minrk, can I work on anything else?

**Comment by sgibson91** [comment]
Thank you for your contribution! Since the contribution period is now over, we will close this issue.
